implement IsaacStateManager

Charleslee522 opened the pull request:
### Github Issue
<!--
    Add the Github issue number if one exists, prefixed by one of Github's keywords, ex. `Fixes #1`, `Closes #1` or `Resolves #1`.
-->
Fixes #251 

### Background
In the new [ISAAC design](https://github.com/bosnet/sebak/wiki/How-the-consensus-protocol-works), a node must have only one IsaacState and all nodes have been designed to have almost same state.

### Solution
Add IsaacState and IsaacStateManager.
IsaacStateManager runs Start() method by goroutine and it manages IsaacState of the node.

### Comment
I wanted to breakdown broadcastor, but I can't because it is deeply related with IsaacState* tests.
I checked #331.
I'm going to rebase after merging #339.

Charleslee522 commented:
Oh, all tests are passed in local but some tests are failed cause of machine performance...
I'll find the way.

spikeekips commented on lib/broadcastor_test.go:
```diff
@@ -0,0 +1,29 @@
+
+	nr := nodeRunners[0]
+
+	b := NewTestBroadcastor()
```
❤️

spikeekips commented on lib/isaac_state_manager.go:
```diff
@@ -0,0 +1,195 @@
+	}
+}
+
+func isTargetLater(current IsaacState, target IsaacState) (result bool) {
```
How about just return when `result` is set without named return.
```go
func isTargetLater(current IsaacState, target IsaacState) bool {
	if current.round.BlockHeight > target.round.BlockHeight {
		return false
	}

	if current.round.BlockHeight < target.round.BlockHeight {
		return true
	}

	// current.round.BlockHeight == target.round.BlockHeight
	if current.round.Number > target.round.Number {
		return false
	}

	if current.round.Number < target.round.Number {
		return true
	}

	// current.round.Number == target.round.Number
	if current.ballotState >= target.ballotState {
		return false
	}

	return true
}
```

anarcher commented on lib/test.go:
```diff
@@ -291,3 +291,18 @@ func ReceiveBallot(t *testing.T, nodeRunner *NodeRunner, ballot *Ballot) error {
+	return p
+}
+
+func (b TestBroadcastor) Broadcast(_ *network.ConnectionManager, message common.Message) (errs map[string]error) {
```
It does not matter but the Broadcaster needs ConnectManager when it broadcast? :-) 

```
type Broadcastor interface {
	Broadcast(common.Message) (errs map[string]error)
}
```

Charleslee522 replied:
@anarcher Is there any problem?? :)

anarcher replied:
@Charleslee522  No, Not really :-) Just MHO :-)

Charleslee522 replied:
@anarcher Oh, I just wanted to know your humble opinion :)

Charleslee522 replied:
In broadcast function, c.connected(address) and c.GetConnection are needed :)
```
func (b SimpleBroadcastor) Broadcast(c *ConnectionManager, message common.Message) (errs map[string]error) {
	for addr, _ := range c.connected {
		go func(v *node.Validator) {
			if v == nil {
				panic("Validator connected but not registered")
			}

			client := c.GetConnection(v.Address())

			var err error
			if message.GetType() == BallotMessage {
				_, err = client.SendBallot(message)
			} else if message.GetType() == string(TransactionMessage) {
				_, err = client.SendMessage(message)
			} else {
				panic("invalid message")
			}

			if err != nil {
				errs[v.Address()] = err
			}
		}(c.validators[addr])
	}
	return
}
```

anarcher replied:
It meant like below: 
```
type Broadcastor interface {
	Broadcast(common.Message) (errs map[string]error)
}

type SimpleBroadcastor struct {
	cm *ConnectionManager
}
func NewSimpleBroadcastor(cm *ConnectionManager) *SimpleBroadcastor

type MultiConnManagerBroadcastor struct { ... }

type GossipBroadcastor struct { ... } 

type StaticBroadcastor 
type TestBroadcastor 

```
But If we need to  choose `ConnectionManager` in  a request scope, We should use  arguments like  `(c *ConnectionManager,msg common.Message) `

Charleslee522 replied:
@anarcher Thank you for giving me a good idea 👍

anarcher commented on lib/isaac_state_manager.go:
```diff
@@ -0,0 +1,195 @@
+	go func() {
+		sm.stop <- true
+	}()
+}
```
What is use-cases that use `Stop`?   If the most case is that node is stopping,  the node/server should wait to stop `IsaacStateManage`?  

IMHO,  It can handle `IsaacStateManger` lifecycle in `main()` with `rungroup` (not node_runner) (It's just another option :-)

spikeekips replied:
@anarcher `Stop` can be used only for testing at this time, but it will he useful :)

Charleslee522 commented:
> Oh, all tests are passed in local but some tests are failed cause of machine performance...
I'll find the way.

I removed sleep and added channels for checking state transition in tests.

Charleslee522 commented:
> How about merge bellow two files into one? 

@kfangw Thank you for the good idea :)

Charleslee522 commented:
@Geod24 @soonkuk Ping~~~~~~~

Geod24 commented on lib/isaac_state.go:
```diff
@@ -0,0 +1,24 @@
+// When a node receive ballot, if the ballotState is former then nodeRunnerState,
```
When a node receive*s* a ballot, if the ballotState is *before* (or in this case, *lower*?) than nodeRunnerState, the ballot is ignored.

Geod24 commented on lib/isaac_state.go:
```diff
@@ -0,0 +1,24 @@
+// When a node receive ballot, if the ballotState is former then nodeRunnerState,
+// the ballot is ignored.
+// The node decides it is former or not by ISAACState.
```
I don't understand this sentence?

Charleslee522 replied:
Sorry, I think it's redundant.

Geod24 commented on lib/isaac_state_manager.go:
```diff
@@ -0,0 +1,209 @@
+	stateTransit  chan ISAACState
+	nextHeight    chan struct{}
+	stop          chan struct{}
+	transitSignal func()
```
Please document what a "transit signal" is.

Geod24 commented on lib/test.go:
```diff
@@ -291,3 +291,23 @@ func ReceiveBallot(t *testing.T, nodeRunner *NodeRunner, ballot *Ballot) error {
+		b.recv <- struct{}{}
+	}
+	return
+}
```
So... That broadcast has no effect anymore ?

Charleslee522 replied:
@Geod24
Which kind of effect did you mean?
FYI, The Broadcaster is used in [TestStateINITTimeoutNotProposer](https://github.com/bosnet/sebak/pull/340/files#diff-cfb80960eeeaa7a01e196aea83c70705R83).

Charleslee522 commented:
@Geod24 I'm done :)

Geod24 commented on lib/broadcastor_test.go:
```diff
@@ -0,0 +1,32 @@
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
```
require

Geod24 commented on lib/isaac_configuration.go:
```diff
@@ -1,8 +1,7 @@
 // The ISAACConfiguration is included in ISAACStateManager and
 // these timeout features are used in ISAAC consensus.
-//
+
```
Unrelated ?

Charleslee522 replied:
Yes 👍 but it is necessary.

Geod24 replied:
Why ?

Charleslee522 replied:
Ah, You want to remove new line, right?
I mean, removing empty comment.

Geod24 replied:
Well this style is used in many places. It allows to create a bit of visual space and not make the code too compact. I'm fine if you want to change that style, we can discuss it, but I just feel like it should be its own PR, since this file is untouched except for this comment change.

Geod24 commented on lib/isaac_state_manager.go:
```diff
@@ -0,0 +1,209 @@
+// ISAACStateManager manages the ISAACState.
+// The most important function `Start()` is called in StartStateManager() function in node_runner.go by goroutine.
```
Ping

Geod24 commented on lib/isaac_state_manager.go:
```diff
@@ -0,0 +1,210 @@
+// ISAACStateManager manages the ISAACState.
+// The most important function `Start()` is called in StartStateManager() function in node_runner.go by goroutine.
+// `transitSignal` is function which is called when the ISAACState is changed.
+// It is used in isaac_state_transit_test.go::TestStateTransitFromTimeoutInitToAccept.
```
Please move this documentation on `transitSignal`.
Also, we're having more and more test-only code... That's not a good sign

Charleslee522 replied:
@Geod24 Yes, I agree. But I don't have another good idea for checking state transit u.u.

Charleslee522 commented:
@Geod24 Thanks for your careful review :) I'm done.
